File: crates/lvz-tools/src/builtins.rs
//! Built-in tools: `read_file`, `write_file`, `list_dir`, and `shell`.
//!
//! These act on the process's working directory. Blast-radius constraints (sandboxing,
//! path allow-lists) are a deliberately later concern (`RECIPE.md` §7.3); for now the tools
//! are thin wrappers over `tokio::fs` / `tokio::process` that report failures back to the
//! model as error results rather than aborting the turn.

use async_trait::async_trait;
use lvz_protocol::{Tool, ToolError, ToolOutput};
use serde::Deserialize;
use serde_json::{json, Value};

/// Parse a tool's argument object into a typed struct, mapping failure to [`ToolError::InvalidArgs`].
fn parse_args<T: for<'de> Deserialize<'de>>(args: Value) -> Result<T, ToolError> {
    serde_json::from_value(args).map_err(|e| ToolError::InvalidArgs(e.to_string()))
}

/// `read_file` — return the UTF-8 contents of a file.
pub struct ReadFileTool;

#[derive(Deserialize)]
struct ReadArgs {
    path: String,
}

#[async_trait]
impl Tool for ReadFileTool {
    fn name(&self) -> &str {
        "read_file"
    }

    fn description(&self) -> &str {
        "Read the UTF-8 contents of a file at the given path (relative to the working directory)."
    }

    fn schema(&self) -> Value {
        json!({
            "type": "object",
            "properties": { "path": { "type": "string", "description": "Path to the file" } },
            "required": ["path"]
        })
    }

    async fn invoke(&self, args: Value) -> Result<ToolOutput, ToolError> {
        let ReadArgs { path } = parse_args(args)?;
        match tokio::fs::read_to_string(&path).await {
            Ok(content) => Ok(ToolOutput::ok(content)),
            Err(e) => Ok(ToolOutput::error(format!("read_file {path}: {e}"))),
        }
    }
}

/// `read_files` — read several files in one round-trip (`RECIPE.md` §6.1 multi-file batching).
pub struct ReadFilesTool;

#[derive(Deserialize)]
struct ReadFilesArgs {
    paths: Vec<String>,
}

/// Concatenate per-file sections under `===== <path> =====` headers; a read failure for one file
/// is inlined as an error section so the rest of the batch still returns.
fn join_sections(sections: impl IntoIterator<Item = (String, String)>) -> String {
    sections
        .into_iter()
        .map(|(path, body)| format!("===== {path} =====\n{body}"))
        .collect::<Vec<_>>()
        .join("\n\n")
}

#[async_trait]
impl Tool for ReadFilesTool {
    fn name(&self) -> &str {
        "read_files"
    }

    fn description(&self) -> &str {
        "Read several files at once and return them concatenated under per-file headers. Prefer \
         this over multiple read_file calls when you need more than one file — it costs one \
         round-trip instead of several. A failure to read one file is reported inline; the rest \
         still return."
    }

    fn schema(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "paths": {
                    "type": "array",
                    "items": { "type": "string" },
                    "description": "Paths to read, relative to the working directory"
                }
            },
            "required": ["paths"]
        })
    }

    async fn invoke(&self, args: Value) -> Result<ToolOutput, ToolError> {
        let ReadFilesArgs { paths } = parse_args(args)?;
        let mut sections = Vec::with_capacity(paths.len());
        for path in paths {
            let body = match tokio::fs::read_to_string(&path).await {
                Ok(content) => content,
                Err(e) => format!("[error: {e}]"),
            };
            sections.push((path, body));
        }
        Ok(ToolOutput::ok(join_sections(sections)))
    }
}

/// `write_file` — create or overwrite a file with the given contents.
pub struct WriteFileTool;

#[derive(Deserialize)]
struct WriteArgs {
    path: String,
    content: String,
}

#[async_trait]
impl Tool for WriteFileTool {
    fn name(&self) -> &str {
        "write_file"
    }

    fn description(&self) -> &str {
        "Create or overwrite a file at the given path with the provided UTF-8 content."
    }

    fn schema(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "path": { "type": "string", "description": "Path to the file" },
                "content": { "type": "string", "description": "Full file contents to write" }
            },
            "required": ["path", "content"]
        })
    }

    async fn invoke(&self, args: Value) -> Result<ToolOutput, ToolError> {
        let WriteArgs { path, content } = parse_args(args)?;
        if let Some(parent) = std::path::Path::new(&path).parent() {
            if !parent.as_os_str().is_empty() {
                let _ = tokio::fs::create_dir_all(parent).await;
            }
        }
        let bytes = content.len();
        // Whether this write actually mutates the workspace: a new file, or content that differs
        // from what's already there. A no-op rewrite (same bytes) is not progress (drives the
        // agent's convergence levers).
        let changed = match tokio::fs::read_to_string(&path).await {
            Ok(existing) => existing != content,
            Err(_) => true,
        };
        match tokio::fs::write(&path, content).await {
            Ok(()) => Ok(ToolOutput::ok(format!("wrote {bytes} bytes to {path}")).changed(changed)),
            Err(e) => Ok(ToolOutput::error(format!("write_file {path}: {e}"))),
        }
    }
}

/// `list_dir` — list the entries of a directory.
pub struct ListDirTool;

#[derive(Deserialize)]
struct ListArgs {
    #[serde(default = "dot")]
    path: String,
}

fn dot() -> String {
    ".".to_string()
}

#[async_trait]
impl Tool for ListDirTool {
    fn name(&self) -> &str {
        "list_dir"
    }

    fn description(&self) -> &str {
        "List the names of entries in a directory (defaults to the working directory). \
         Directory entries are suffixed with '/'."
    }

    fn schema(&self) -> Value {
        json!({
            "type": "object",
            "properties": { "path": { "type": "string", "description": "Directory path (default '.')" } }
        })
    }

    async fn invoke(&self, args: Value) -> Result<ToolOutput, ToolError> {
        let ListArgs { path } = parse_args(args)?;
        let mut entries = match tokio::fs::read_dir(&path).await {
            Ok(rd) => rd,
            Err(e) => return Ok(ToolOutput::error(format!("list_dir {path}: {e}"))),
        };
        let mut names = Vec::new();
        loop {
            match entries.next_entry().await {
                Ok(Some(entry)) => {
                    let name = entry.file_name().to_string_lossy().into_owned();
                    let is_dir = entry.file_type().await.map(|t| t.is_dir()).unwrap_or(false);
                    names.push(if is_dir { format!("{name}/") } else { name });
                }
                Ok(None) => break,
                Err(e) => return Ok(ToolOutput::error(format!("list_dir {path}: {e}"))),
            }
        }
        names.sort();
        Ok(ToolOutput::ok(names.join("\n")))
    }
}

/// `shell` — run a command line via `sh -c` and return combined output + exit status.
pub struct ShellTool;

#[derive(Deserialize)]
struct ShellArgs {
    command: String,
}

#[async_trait]
impl Tool for ShellTool {
    fn name(&self) -> &str {
        "shell"
    }

    fn description(&self) -> &str {
        "Run a shell command via `sh -c` in the working directory and return its stdout, \
         stderr, and exit code."
    }

    fn schema(&self) -> Value {
        json!({
            "type": "object",
            "properties": { "command": { "type": "string", "description": "Command line to execute" } },
            "required": ["command"]
        })
    }

    async fn invoke(&self, args: Value) -> Result<ToolOutput, ToolError> {
        let ShellArgs { command } = parse_args(args)?;
        let output = tokio::process::Command::new("sh")
            .arg("-c")
            .arg(&command)
            .output()
            .await
            .map_err(|e| ToolError::Execution(format!("spawning `{command}`: {e}")))?;

        let code = output.status.code().unwrap_or(-1);
        let stdout = String::from_utf8_lossy(&output.stdout);
        let stderr = String::from_utf8_lossy(&output.stderr);
        let rendered = format!("exit={code}\n--- stdout ---\n{stdout}\n--- stderr ---\n{stderr}");
        // A non-zero exit is reported as an error result so the model can react, but it is
        // not a hard ToolError — the tool itself ran fine.
        Ok(ToolOutput {
            content: rendered,
            is_error: !output.status.success(),
            // The shell is not one of the agent's tracked edit tools, so its mutation (if any) is
            // not used as a convergence signal; leave `changed` false.
            changed: false,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn read_missing_file_is_a_soft_error() {
        let out = ReadFileTool
            .invoke(json!({ "path": "/nonexistent/lvz/xyz" }))
            .await
            .unwrap();
        assert!(out.is_error);
    }

    #[tokio::test]
    async fn write_then_read_round_trips() {
        let dir = std::env::temp_dir().join(format!("lvz_tools_{}", std::process::id()));
        let path = dir.join("hello.txt");
        let path_str = path.to_string_lossy().to_string();

        let w = WriteFileTool
            .invoke(json!({ "path": path_str, "content": "hi there" }))
            .await
            .unwrap();
        assert!(!w.is_error, "write failed: {}", w.content);

        let r = ReadFileTool
            .invoke(json!({ "path": path_str }))
            .await
            .unwrap();
        assert_eq!(r.content, "hi there");

        let _ = tokio::fs::remove_dir_all(&dir).await;
    }

    #[tokio::test]
    async fn shell_reports_exit_code_and_output() {
        let ok = ShellTool
            .invoke(json!({ "command": "echo lavoisier" }))
            .await
            .unwrap();
        assert!(!ok.is_error);
        assert!(ok.content.contains("lavoisier"));
        assert!(ok.content.contains("exit=0"));

        let bad = ShellTool
            .invoke(json!({ "command": "exit 3" }))
            .await
            .unwrap();
        assert!(bad.is_error);
        assert!(bad.content.contains("exit=3"));
    }

    #[tokio::test]
    async fn invalid_args_are_rejected() {
        let err = ReadFileTool.invoke(json!({})).await.unwrap_err();
        assert!(matches!(err, ToolError::InvalidArgs(_)));
    }
}

File: src/Lavoisier/Tool/Builtins.hs
-- | The built-in tools: filesystem reads/writes, directory listing, and shell. Ported from Rust
-- @lvz-tools@ @builtins.rs@. Each is a 'Tool' record (the @Arc\<dyn Tool\>@ analogue). Tool-visible
-- failures (missing file, non-zero exit) are 'toolErr' results (the turn continues); only a bad
-- argument shape is a hard 'ToolError'.
module Lavoisier.Tool.Builtins
  ( builtinTools,
    readFileTool,
    readFilesTool,
    writeFileTool,
    listDirTool,
    shellTool,
    outlineFileTool,
    outlineFilesTool,
  )
where

import Control.Exception (IOException, try)
import Data.Aeson (Value (..), object, (.=))
import Data.Aeson.Key qualified as K
import Data.Aeson.KeyMap qualified as KM
import Data.ByteString qualified as BS
import Data.ByteString.Lazy qualified as BL
import Data.List (sort)
import Data.Map.Strict qualified as Map
import Data.Scientific (toBoundedInteger)
import Data.Set qualified as Set
import Data.Text (Text)
import Data.Text qualified as T
import Data.Text.Encoding (decodeUtf8Lenient, encodeUtf8)
import Data.Text.IO qualified as TIO
import Lavoisier.Context.Lang (langFromPath)
import Lavoisier.Context.Skeleton qualified as Skel
import Lavoisier.Context.Symbols qualified as Sym
import Lavoisier.Protocol.Tool
import System.Directory
  ( createDirectoryIfMissing,
    doesDirectoryExist,
    listDirectory,
  )
import System.Exit (ExitCode (..))
import System.FilePath (takeDirectory)
import System.Process.Typed (proc, readProcess)

-- | The default built-in tool set, in registration order.
builtinTools :: [Tool]
builtinTools =
  [ readFileTool,
    readFilesTool,
    writeFileTool,
    listDirTool,
    shellTool,
    outlineFileTool,
    outlineFilesTool
  ]

readFileTool :: Tool
readFileTool =
  Tool
    { toolName = "read_file",
      toolDescription =
        "Read the UTF-8 contents of a file at the given path (relative to the working directory).",
      toolSchema =
        object
          [ "type" .= sv "object",
            "properties" .= object ["path" .= prop "Path to the file"],
            "required" .= [sv "path"]
          ],
      toolInvoke = \args -> withArg (argStr "path" args) $ \path -> do
        r <- tryIO (TIO.readFile (T.unpack path))
        pure . Right $ case r of
          Right content -> toolOk content
          Left e -> toolErr ("read_file " <> path <> ": " <> tshow e)
    }

readFilesTool :: Tool
readFilesTool =
  Tool
    { toolName = "read_files",
      toolDescription =
        "Read several files at once and return them concatenated under per-file headers. Prefer "
          <> "this over multiple read_file calls when you need more than one file. A failure to read "
          <> "one file is reported inline; the rest still return.",
      toolSchema =
        object
          [ "type" .= sv "object",
            "properties"
              .= object
                [ "paths"
                    .= object
                      [ "type" .= sv "array",
                        "items" .= object ["type" .= sv "string"],
                        "description" .= sv "Paths to read, relative to the working directory"
                      ]
                ],
            "required" .= [sv "paths"]
          ],
      toolInvoke = \args -> withArg (argStrList "paths" args) $ \paths -> do
        sections <- mapM readOne paths
        pure (Right (toolOk (T.intercalate "\n\n" sections)))
    }
  where
    readOne path = do
      r <- tryIO (TIO.readFile (T.unpack path))
      let body = either (\e -> "[error: " <> tshow e <> "]") id r
      pure ("===== " <> path <> " =====\n" <> body)

writeFileTool :: Tool
writeFileTool =
  Tool
    { toolName = "write_file",
      toolDescription = "Create or overwrite a file at the given path with the provided UTF-8 content.",
      toolSchema =
        object
          [ "type" .= sv "object",
            "properties"
              .= object
                [ "path" .= prop "Path to the file",
                  "content" .= prop "Full file contents to write"
                ],
            "required" .= [sv "path", sv "content"]
          ],
      toolInvoke = \args -> withArg (both (argStr "path" args) (argStr "content" args)) $ \(path, content) -> do
        let dir = takeDirectory (T.unpack path)
        _ <- tryIO (createDirectoryIfMissing True dir)
        existing <- tryIO (TIO.readFile (T.unpack path))
        let changed = either (const True) (/= content) existing
            nbytes = BS.length (encodeUtf8 content)
        w <- tryIO (TIO.writeFile (T.unpack path) content)
        pure . Right $ case w of
          Right () -> setChanged changed (toolOk ("wrote " <> tshow nbytes <> " bytes to " <> path))
          Left e -> toolErr ("write_file " <> path <> ": " <> tshow e)
    }

listDirTool :: Tool
listDirTool =
  Tool
    { toolName = "list_dir",
      toolDescription =
        "List the names of entries in a directory (defaults to the working directory). "
          <> "Directory entries are suffixed with '/'.",
      toolSchema =
        object
          [ "type" .= sv "object",
            "properties" .= object ["path" .= prop "Directory path (default '.')"]
          ],
      toolInvoke = \args -> withArg (Right (argStrOr "path" "." args)) $ \path -> do
        r <- tryIO (listDirectory (T.unpack path))
        case r of
          Left e -> pure (Right (toolErr ("list_dir " <> path <> ": " <> tshow e)))
          Right names -> do
            decorated <- mapM (decorate path) names
            pure (Right (toolOk (T.intercalate "\n" (sort decorated))))
    }
  where
    decorate path name = do
      isDir <- doesDirectoryExist (T.unpack path <> "/" <> name)
      pure (T.pack (if isDir then name <> "/" else name))

shellTool :: Tool
shellTool =
  Tool
    { toolName = "shell",
      toolDescription =
        "Run a shell command via `sh -c` in the working directory and return its stdout, stderr, "
          <> "and exit code.",
      toolSchema =
        object
          [ "type" .= sv "object",
            "properties" .= object ["command" .= prop "Command line to execute"],
            "required" .= [sv "command"]
          ],
      toolInvoke = \args -> withArg (argStr "command" args) $ \command -> do
        r <- tryIO (readProcess (proc "sh" ["-c", T.unpack command]))
        pure . Right $ case r of
          Left e -> toolErr ("shell: " <> tshow e)
          Right (code, out, err) ->
            let n = case code of ExitSuccess -> 0 :: Int; ExitFailure k -> k
                rendered =
                  "exit="
                    <> tshow n
                    <> "\n--- stdout ---\n"
                    <> decodeUtf8Lenient (BL.toStrict out)
                    <> "\n--- stderr ---\n"
                    <> decodeUtf8Lenient (BL.toStrict err)
             in (if code == ExitSuccess then toolOk else toolErr) rendered
    }

-- --- context-engine tools (skeleton + symbol references over tree-sitter) -------------------------

-- | Skeletonise one source file (signatures kept, bodies elided), with an optional symbol focus.
outlineFileTool :: Tool
outlineFileTool =
  Tool
    { toolName = "outline_file",
      toolDescription =
        "Return a token-efficient skeleton of a source file: signatures, type definitions and doc "
          <> "comments kept, function/method bodies elided. Optionally focus on a symbol — pass "
          <> "`focus` (and `radius`, default 1) to keep full bodies for it and everything within "
          <> "`radius` dependency hops. Supported: Rust, Python, JavaScript, TypeScript; other files "
          <> "are returned unchanged. Prefer this over read_file when you only need a file's shape.",
      toolSchema =
        object
          [ "type" .= sv "object",
            "properties"
              .= object
                [ "path" .= prop "Path to the source file",
                  "focus" .= prop "Optional symbol to expand around (keep its body + dependencies)",
                  "radius"
                    .= object
                      [ "type" .= sv "integer",
                        "minimum" .= (0 :: Int),
                        "description" .= sv "Dependency-hop radius for focus (default 1)"
                      ]
                ],
            "required" .= [sv "path"]
          ],
      toolInvoke = \args -> withArg (argStr "path" args) $ \path -> do
        r <- tryIO (BS.readFile (T.unpack path))
        case r of
          Left e -> pure (Right (toolErr ("outline_file " <> path <> ": " <> tshow e)))
          Right bytes -> case langFromPath (T.unpack path) of
            Nothing -> pure (Right (toolOk (decodeUtf8Lenient bytes)))
            Just lang -> do
              out <- case argStr "focus" args of
                Right focus -> Sym.skeletonWithRadius lang focus (max 0 (argIntOr "radius" 1 args)) bytes
                Left _ -> Skel.skeleton lang bytes
              pure (Right (toolOk (decodeUtf8Lenient out)))
    }

-- | Skeletonise several source files at once, concatenated under per-file headers.
outlineFilesTool :: Tool
outlineFilesTool =
  Tool
    { toolName = "outline_files",
      toolDescription =
        "Skeletonise several source files at once (signatures kept, bodies elided), concatenated "
          <> "under per-file headers. Prefer this over multiple outline_file calls. With `focus`, "
          <> "bodies are kept for symbols within `radius` dependency hops of it — followed ACROSS the "
          <> "given files, so pass every file that might be involved. Unsupported file types are "
          <> "returned unchanged.",
      toolSchema =
        object
          [ "type" .= sv "object",
            "properties"
              .= object
                [ "paths"
                    .= object
                      [ "type" .= sv "array",
                        "items" .= object ["type" .= sv "string"],
                        "description" .= sv "Paths to outline, relative to the working directory"
                      ],
                  "focus" .= prop "Optional symbol to expand around in each file",
                  "radius"
                    .= object
                      [ "type" .= sv "integer",
                        "minimum" .= (0 :: Int),
                        "description" .= sv "Dependency-hop radius for focus (default 1)"
                      ]
                ],
            "required" .= [sv "paths"]
          ],
      toolInvoke = \args -> withArg (argStrList "paths" args) $ \paths -> do
        let mfocus = either (const Nothing) Just (argStr "focus" args)
            radius = max 0 (argIntOr "radius" 1 args)
        entries <- mapM (\p -> (,) p <$> load p) paths
        bodies <- renderAll mfocus radius entries
        pure (Right (toolOk (T.intercalate "\n\n" (zipWith section paths bodies))))
    }
  where
    -- Read a path into what the skeletoniser needs, or the text to print in its place.
    load path = do
      r <- tryIO (BS.readFile (T.unpack path))
      pure $ case r of
        Left e -> Left ("[error: " <> tshow e <> "]")
        Right bytes -> case langFromPath (T.unpack path) of
          Nothing -> Left (decodeUtf8Lenient bytes)
          Just lang -> Right (lang, bytes)

    -- Without a focus each file is skeletonised on its own. With one, the dependency graph spans the
    -- whole set, so a caller in one file keeps the body of the callee it reaches in another — which
    -- is the point of passing several paths, and where the import-ranked cross-file edges earn their
    -- keep.
    renderAll Nothing _ entries =
      mapM (either pure (\(lang, bytes) -> decodeUtf8Lenient <$> Skel.skeleton lang bytes) . snd) entries
    renderAll (Just focus) radius entries = do
      let parsable = [(i, path, lang, bytes) | (i, (path, Right (lang, bytes))) <- zip [0 :: Int ..] entries]
      graph <- Sym.fromFiles [Sym.SourceFile p l b | (_, p, l, b) <- parsable]
      let byIndex =
            Map.fromList
              (zip [i | (i, _, _, _) <- parsable] (Sym.neighborsWithinByFile focus radius graph))
      mapM
        ( \(i, (_, e)) -> case e of
            Left msg -> pure msg
            Right (lang, bytes) ->
              decodeUtf8Lenient <$> Skel.skeletonize lang (Map.findWithDefault Set.empty i byIndex) bytes
        )
        (zip [0 :: Int ..] entries)

    section path body = "===== " <> path <> " =====\n" <> body

-- --- argument parsing + small helpers -------------------------------------------------------------

-- | Run an IO effect only when the arguments parsed; a parse failure short-circuits to 'ToolError'.
withArg :: Either ToolError a -> (a -> IO (Either ToolError ToolOutput)) -> IO (Either ToolError ToolOutput)
withArg (Left e) _ = pure (Left e)
withArg (Right a) k = k a

argStr :: Text -> Value -> Either ToolError Text
argStr key (Object o) = case KM.lookup (K.fromText key) o of
  Just (String s) -> Right s
  Just _ -> Left (TEInvalidArgs (key <> " must be a string"))
  Nothing -> Left (TEInvalidArgs ("missing argument: " <> key))
argStr _ _ = Left (TEInvalidArgs "expected a JSON object")

argStrOr :: Text -> Text -> Value -> Text
argStrOr key def v = either (const def) id (argStr key v)

-- | An integer argument with a default (a non-integer or absent value falls back to @def@).
argIntOr :: Text -> Int -> Value -> Int
argIntOr key def (Object o) = case KM.lookup (K.fromText key) o of
  Just (Number n) -> maybe def id (toBoundedInteger n)
  _ -> def
argIntOr _ def _ = def

argStrList :: Text -> Value -> Either ToolError [Text]
argStrList key (Object o) = case KM.lookup (K.fromText key) o of
  Just (Array a) -> traverse asStr (foldr (:) [] a)
  Just _ -> Left (TEInvalidArgs (key <> " must be an array"))
  Nothing -> Left (TEInvalidArgs ("missing argument: " <> key))
  where
    asStr (String s) = Right s
    asStr _ = Left (TEInvalidArgs (key <> " must be an array of strings"))
argStrList _ _ = Left (TEInvalidArgs "expected a JSON object")

both :: Either e a -> Either e b -> Either e (a, b)
both ea eb = (,) <$> ea <*> eb

tryIO :: IO a -> IO (Either IOException a)
tryIO = try

sv :: Text -> Value
sv = String

-- | A @{"type":"string","description":…}@ schema property.
prop :: Text -> Value
prop desc = object ["type" .= sv "string", "description" .= sv desc]

tshow :: (Show a) => a -> Text
tshow = T.pack . show
